In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_excel('dados_vendas.xlsx')

In [ ]:
compras=df.COD_CLIENTE.value_counts().to_frame()
#compras[compras['COD_CLIENTE']>1].describe()
#compras.describe()

In [ ]:
n=len(compras)
slices=10
print("%\tCount\tMin\tmean\tstd")
for x in range(1,slices):
    percentage=(slices-x)*(100/slices)
    data=compras[:int(-n*(x/slices))]['COD_CLIENTE']
    print(str(percentage)+"%",data.count(),data.min(),"%.2f"%data.mean(),"%.2f"%data.std(),sep='\t')

In [136]:
vendas=df.COD_PRODUTO.value_counts().to_frame()
n=len(vendas)
slices=10
print("%\tCount\tMin\tmean\tstd")
for x in range(1,slices):
    percentage=(slices-x)*(100/slices)
    data=vendas[:int(-n*(x/slices))]['COD_PRODUTO']
    print(str(percentage)+"%",data.count(),data.min(),"%.2f"%data.mean(),"%.2f"%data.std(),sep='\t')

%	Count	Min	mean	std
90.0%	644	1	499.49	1348.54
80.0%	572	1	562.23	1418.66
70.0%	501	2	641.70	1499.13
60.0%	429	4	748.98	1595.36
50.0%	358	14	896.09	1708.87
40.0%	286	61	1113.49	1849.90
30.0%	215	198	1442.33	2029.60
20.0%	143	509	2001.41	2294.39
10.0%	72	1093	3236.38	2716.92


In [139]:
vendas=df.COD_PRODUTO.value_counts().to_frame()
print(vendas[:58])

       COD_PRODUTO
26243        14001
21653        12909
25727        10397
25729        10264
25726         8090
25981         7633
25996         6935
25982         6855
26222         6160
25983         5675
25354         5239
25728         5142
25172         4662
25481         4620
21647         4318
25545         4105
25303         4011
23802         3912
26005         3788
26004         3560
25676         3429
26003         3374
21255         3369
25173         3106
26324         3032
19729         3020
25355         2994
25998         2951
25546         2777
25997         2752
21854         2716
26247         2686
21256         2598
21855         2431
24913         2429
25870         2420
24716         2409
25366         2322
25933         2198
22752         2157
25730         2096
26423         1957
25577         1805
26325         1745
26007         1705
14537         1703
20966         1651
26387         1615
25904         1609
25675         1608
26008         1592
25193       

In [131]:
vendas=df.COD_PRODUTO.value_counts().to_frame()
first=np.percentile(vendas, 25)  
third=np.percentile(vendas, 75)
IQR=third-first
print("IQR: ",IQR)
outlier_threshold=3*IQR+third
n_antigo=len(vendas)
vendas=vendas[vendas['COD_PRODUTO']<outlier_threshold]
n=len(vendas)
print(n_antigo-n," Outliers excluidos")
slices=10
print("%\tCount\tMin\tmean\tstd")
for x in range(1,slices):
    percentage=(slices-x)*(100/slices)
    data=vendas[:int(-n*(x/slices))]['COD_PRODUTO']
    print(str(percentage)+"%",data.count(),data.min(),"%.2f"%data.mean(),"%.2f"%data.std(),sep='\t')

IQR:  321.0
58  Outliers excluidos
%	Count	Min	mean	std
90.0%	592	1	177.45	297.60
80.0%	526	1	199.59	308.70
70.0%	460	2	228.05	320.20
60.0%	395	3	265.22	331.12
50.0%	329	8	317.51	339.54
40.0%	263	33	393.00	340.30
30.0%	198	102	502.62	324.06
20.0%	132	286	662.33	281.71
10.0%	66	607	896.53	197.75


In [24]:
for x in range(10):
    teste= pd.qcut(df.COD_PRODUTO,25)
    break
teste

0         (25727.0, 25729.0]
1         (24851.0, 25189.0]
2         (26178.0, 26243.0]
3         (26178.0, 26243.0]
4         (26572.0, 26727.0]
                 ...        
321736    (23802.0, 24851.0]
321737    (25869.0, 25981.0]
321738    (25869.0, 25981.0]
321739    (25718.0, 25727.0]
321740    (25718.0, 25727.0]
Name: COD_PRODUTO, Length: 321741, dtype: category
Categories (25, interval[float64]): [(10215.999, 21029.0] < (21029.0, 21653.0] < (21653.0, 21821.0] < (21821.0, 22617.0] ... (26243.0, 26324.0] < (26324.0, 26424.0] < (26424.0, 26572.0] < (26572.0, 26727.0]]

Número de clientes = 236244 <br>
Número de produtos = 715  <br>
Número de compras  = 321741 <br>
Número de compras sem repetição = 311741 <br>
Número de compradores com mais de uma compra= 57365 <br>

In [21]:
newdf= df[['COD_CLIENTE','COD_PRODUTO']]
threshold=5
newdf= newdf[newdf['COD_CLIENTE'].map(newdf['COD_CLIENTE'].value_counts())>=threshold ]
v=newdf['COD_CLIENTE'].nunique()
u=newdf['COD_PRODUTO'].nunique()

convert_cliente={}
revert_cliente=['' for x in range(v)]
c=0
for x in newdf['COD_CLIENTE'].unique():
    convert_cliente[x]=c
    revert_cliente[c]=x
    c+=1
convert_produto={}
revert_produto=['' for x in range(u)]
c=0
for x in newdf['COD_PRODUTO'].unique():
    convert_produto[x]=c
    revert_produto[c]=x
    c+=1
compras_unicas=0
lista_adj=[[] for x in range(v)]
for i,j in newdf.iterrows():
    if(convert_produto[j[1]] not in lista_adj[convert_cliente[j[0]]]):
        lista_adj[convert_cliente[j[0]]].append(convert_produto[j[1]])
        compras_unicas+=1
duplicados=[]
for j in range(len(lista_adj)):
    if(len(lista_adj[j])<threshold):
        duplicados.append(j)
for index in sorted(duplicados, reverse=True):
    del lista_adj[index]
v=len(lista_adj)

teste=[x for l in lista_adj for x in l]
pd.set_option('display.max_rows', 500)
produtos_tratado=pd.DataFrame.from_dict({x:teste.count(x) for x in teste}, orient='index').sort_values(0,ascending=False)
#print(produtos_tratado.describe())
u=len(set(x for l in lista_adj for x in l))
outliers_u=produtos_tratado[produtos_tratado>=200].dropna()
u_excluded=outliers_u.count()


print(outliers_u.index)
for index in sorted(outliers_u.index, reverse=True):
    for i in range(len(lista_adj)):
        lista_adj[i]=list(filter(lambda a: a not in  outliers_u, lista_adj[i]))
        
        
        
outputname="Adjacency_list_v>5_u<200.txt"
text_file = open(outputname, "w")
text_file.write(str(v)+'\n')
text_file.write(str(u)+'\n')
for j in lista_adj:
    if(len(j)<threshold):
        continue
    txt=str(j)[1:-1].replace(',',' ')
    text_file.write( txt+'\n')
text_file.close()

#p = Popen(['./AMRPB',outputname], shell=True, stdout=PIPE, stdin=PIPE)


Int64Index([0, 26, 32, 23, 33, 7], dtype='int64')


In [6]:
newdf= df[['COD_CLIENTE','COD_PRODUTO']]
threshold_v=5
threshold_u=200
newdf=newdf.drop_duplicates(keep='first')
print(newdf.shape)
newdf= newdf[newdf['COD_CLIENTE'].map(newdf['COD_CLIENTE'].value_counts())>=threshold_v ]
print(newdf.shape)
newdf= newdf[newdf['COD_PRODUTO'].map(newdf['COD_PRODUTO'].value_counts())<=threshold_u ]
print(newdf.shape)
v=newdf['COD_CLIENTE'].nunique()
u=newdf['COD_PRODUTO'].nunique()

pd.set_option('display.max_rows', 500)

convert_cliente={}
revert_cliente=['' for x in range(v)]
c=0
for x in newdf['COD_CLIENTE'].unique():
    convert_cliente[x]=c
    revert_cliente[c]=x
    c+=1
convert_produto={}
revert_produto=['' for x in range(u)]
c=0
for x in newdf['COD_PRODUTO'].unique():
    convert_produto[x]=c
    revert_produto[c]=x
    c+=1
lista_adj=[[] for x in range(v)]
for i,j in newdf.iterrows():
    lista_adj[convert_cliente[j[0]]].append(convert_produto[j[1]])
teste=[x for l in lista_adj for x in l]
pd.set_option('display.max_rows', 500)
produtos_tratado=pd.DataFrame.from_dict({x:teste.count(x) for x in teste}, orient='index').sort_values(0,ascending=False)
print(produtos_tratado.describe())
outputname="Adjacency_list_v>"+str(threshold_v)+"_u<"+str(threshold_u)+".txt"
text_file = open(outputname, "w")
text_file.write(str(v)+'\n')
text_file.write(str(u)+'\n')
for j in lista_adj:
    txt=str(j)[1:-1].replace(',',' ')
    text_file.write( txt+'\n')
text_file.close()

#p = Popen(['./AMRPB',outputname], shell=True, stdout=PIPE, stdin=PIPE)


(311741, 2)
(10226, 2)
(8404, 2)
                0
count  407.000000
mean    20.648649
std     31.635344
min      1.000000
25%      2.000000
50%      8.000000
75%     26.000000
max    181.000000


In [90]:
f = open("saida.txt", "r")
nonnumbers=['(',')','}',']']
text=f.read().split('[')
biclusters=[]
v_cluster=pd.DataFrame(columns=['code','cluster'])
u_cluster=pd.DataFrame(columns=['code','cluster'])
for i,a in enumerate(text[1:]):
    bicluster=[]
    for j,b in enumerate(a.split('),')):
        array=[]
        for k,c in enumerate(b.replace('],','').split(',')):
            txt=c
            for r in nonnumbers:
                txt=txt.replace(r, "")
            if(txt!=''):
                code=int(txt)
                array.append(code)
                new_row = {'code':code, 'cluster':i}
                if(j):
                    v_cluster=v_cluster.append(new_row,ignore_index=True)
                else:
                    u_cluster=u_cluster.append(new_row,ignore_index=True)
            #print(i,j,k,txt,sep='\t')
        bicluster.append(array)
    biclusters.append(bicluster)
print(u_cluster)

      code cluster
0      281       0
1      108       1
2      127       2
3      142       3
4       49       3
...    ...     ...
2329   634     835
2330   635     835
2331   710     835
2332   794     835
2333  1645     837

[2334 rows x 2 columns]
